### PR TITLE
bidViewablityIO Module: add new submodule for detecting viewability without ad server dependancies

### PR DESCRIPTION
**Pull request by jsut**
## Type of change

- [x] Feature

## Description of change

Emits a BID_VIEWABLE event for banner ads when a bid meets IAB viewable specifications, using the browsers IntersectionObserver API, if it is available.

To see this in action locally, gulp serve, then browse to the included integration example.

I haven't got a docs repo PR yet, I'll put that together soon.

## Other information

related issue: https://github.com/prebid/Prebid.js/issues/6909

### CODE REVIEW

File: modules/bidViewabilityIO.js
Comment (robertrmartinez) [inline modules/bidViewabilityIO.js:76]: I think it would be better to never register for this event until the user enables the config.

So your init function could subscribe to setConfig and then once it is enabled, it then will add the event listener.

```js
config.getConfig(MODULE_NAME, conf => {
   // if conf enabled then 
   events.on(EVENTS.AD_RENDER_SUCCEEDED, ({doc, bid, id}) => {
});
```

That way the event callback is not executing every single time this module is included.

Only when it is actually turned on.

I would also add the `CLIENT_SUPPORTS_IO` check inside of that config subscribe as well since no need to run it when it is not supported

**Review by robertrmartinez — CHANGES_REQUESTED** [review]

File: modules/bidViewabilityIO.js
Comment (jsut) [reply modules/bidViewabilityIO.js]: i just pushed a change that doesn't do the subscribe thing, then actually read all these words and found out that this subscribe thing is a thing.  I'll have a look it doing it as you describe this afternoon probably.

Comment (robertrmartinez) [reply modules/bidViewabilityIO.js]: Ok cool.

So basically if you do the following

```js
config.getConfig(MODULE_NAME, CALLBACK_FUNCTION);
```

Prebid will execute your `CALLBACK_FUNCTION` each time `pbjs.setConfig` is called with your `MODULE_NAME`

So for your use case, we only want to register the event listener if the user / pub runs pbjs.setConfig and turns on bidViewability.

so your callback can do something like:

```js
config.getConfig(MODULE_NAME, conf => {
   const  globalModuleConfig = conf.bidViewabilityIO;
   if (globalModuleConfig[CONFIG_ENABLED] && CLIENT_SUPPORTS_IO) {
      // register your event code
      events.on(EVENTS.AD_RENDER_SUCCEEDED, ({doc, bid, id}) => {
   }
});
```

File: modules/bidViewabilityIO.js
Comment (robertrmartinez) [inline modules/bidViewabilityIO.js:23]: This is causing tests in IE to fail becuase array.includes is not supported.

Can you use indexOf instead please.

**Review by robertrmartinez — CHANGES_REQUESTED** [review]

File: modules/bidViewabilityIO.js
Comment (robertrmartinez) [inline modules/bidViewabilityIO.js:42]: This is me being picky, but when I am testing / debugging pub pages I like when modules emit their name inside their console logs, makes it easy to filter.

You could wrap logMessage in your own function which just adds in MODULE_NAME on each message you send.

Comment (jsut) [reply modules/bidViewabilityIO.js:42]: That doesn't sound picky, it sounds like an excellent practice.

**Review by robertrmartinez — APPROVED** [review]

**Comment by robertrmartinez** [comment]
I think we need a docs pr before merging this

**Comment by jsut** [comment]
yup, i'll get on that tomorrow @robertrmartinez

**Comment by jsut** [comment]
docs PR: https://github.com/prebid/prebid.github.io/pull/3184

**Comment by robertrmartinez** [comment]
Nice thanks @jsut 

@ChrisHuie Would you mind giving this a second look? Looks good to me but would love a second pair of eyes before merging this week!

**Review by ChrisHuie — APPROVED** [review]

**Comment by ChrisHuie** [comment]
LGTM. Thanks @jsut for this awesome pr 🙏

**Pull request merged**
